File: 2021/day03/python/solution.py
#!/usr/bin/env python3
import os

def parse_input():
    input_path = os.path.join(os.path.dirname(__file__), '..', 'input.txt')
    with open(input_path) as f:
        return [line.strip() for line in f if line.strip()]

def part1(numbers):
    num_bits = len(numbers[0])
    gamma = 0

    for pos in range(num_bits):
        ones = sum(1 for n in numbers if n[pos] == '1')
        zeros = len(numbers) - ones

        if ones >= zeros:
            gamma |= (1 << (num_bits - 1 - pos))

    # epsilon is bitwise NOT of gamma (within num_bits)
    epsilon = gamma ^ ((1 << num_bits) - 1)

    return gamma * epsilon

def find_rating(numbers, use_most_common):
    num_bits = len(numbers[0])
    candidates = numbers[:]

    for pos in range(num_bits):
        if len(candidates) == 1:
            break

        ones = sum(1 for n in candidates if n[pos] == '1')
        zeros = len(candidates) - ones

        if use_most_common:
            target = '1' if ones >= zeros else '0'
        else:
            target = '0' if zeros <= ones else '1'

        candidates = [n for n in candidates if n[pos] == target]

    return int(candidates[0], 2)

def part2(numbers):
    oxygen = find_rating(numbers, use_most_common=True)
    co2 = find_rating(numbers, use_most_common=False)
    return oxygen * co2

if __name__ == '__main__':
    numbers = parse_input()
    print(f"Part 1: {part1(numbers)}")
    print(f"Part 2: {part2(numbers)}")

File: 2021/day03/ruby/solution.rb
#!/usr/bin/env ruby

def parse_input
  input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
  File.readlines(input_path).map(&:strip).reject(&:empty?)
end

def part1(numbers)
  num_bits = numbers[0].length
  gamma = 0

  num_bits.times do |pos|
    ones = numbers.count { |n| n[pos] == '1' }
    zeros = numbers.length - ones

    gamma |= (1 << (num_bits - 1 - pos)) if ones >= zeros
  end

  # epsilon is bitwise NOT of gamma (within num_bits)
  epsilon = gamma ^ ((1 << num_bits) - 1)

  gamma * epsilon
end

def find_rating(numbers, use_most_common)
  num_bits = numbers[0].length
  candidates = numbers.dup

  num_bits.times do |pos|
    break if candidates.length == 1

    ones = candidates.count { |n| n[pos] == '1' }
    zeros = candidates.length - ones

    target = if use_most_common
               ones >= zeros ? '1' : '0'
             else
               zeros <= ones ? '0' : '1'
             end

    candidates = candidates.select { |n| n[pos] == target }
  end

  candidates[0].to_i(2)
end

def part2(numbers)
  oxygen = find_rating(numbers, true)
  co2 = find_rating(numbers, false)
  oxygen * co2
end

if __FILE__ == $PROGRAM_NAME
  numbers = parse_input
  puts "Part 1: #{part1(numbers)}"
  puts "Part 2: #{part2(numbers)}"
end
